Mathlib/CategoryTheory/Sites/NonabelianCohomology/H1.lean

```diff
@@ -0,0 +1,213 @@
+which is trivial iff `x₃` can be lifted to a global section of `G₂`.
+(This should hold more generally if `G₂` is a sheaf of sets on which `G₁` acts
+freely, and `G₃` is the quotient sheaf.)
+* deduce a similar result for abelian sheaves
+* when the notion of quasi-coherent sheaves on schemes is defined, show that
+if `0 ⟶ Q ⟶ M ⟶ N ⟶ 0` is an exact sequence of abelian sheaves over a scheme `X`
+and `Q` is the underlying sheaf of a quasi-coherent sheaf, then `M(U) ⟶ N(U)`
+is surjective for any affine open `U`.
+* take the colimit of `OneCohomology G U` over all covering families `U` (for
+a Grothendieck topology)
+
+# References
+
+* [J. Frenkel, *Cohomologie non abélienne et espaces fibrés*][frenkel1957]
+
+-/
+
+universe w' w v u
+
+namespace CategoryTheory
+
+variable {C : Type u} [Category.{v} C]
+
+namespace PresheafOfGroups
+
+variable (G : Cᵒᵖ ⥤ GroupCat.{w}) {X : C} {I : Type w'} (U : I → C)
+
+/-- A zero cochain consists of a family of sections. -/
+def ZeroCochain := ∀ (i : I), G.obj (Opposite.op (U i))
+
+instance : Group (ZeroCochain G U) := by
+  dsimp [ZeroCochain]
+  infer_instance
+
+namespace Cochain₀
+
+@[simp]
+lemma one_apply (i : I) : (1 : ZeroCochain G U) i = 1 := rfl
+
+@[simp]
+lemma inv_apply (γ : ZeroCochain G U) (i : I) : γ⁻¹ i = (γ i)⁻¹ := rfl
+
+@[simp]
+lemma mul_apply (γ₁ γ₂ : ZeroCochain G U) (i : I) : (γ₁ * γ₂) i = γ₁ i * γ₂ i := rfl
+
+end Cochain₀
+
+/-- A 1-cochain of a presheaf of groups `G : Cᵒᵖ ⥤ GroupCat` on a family `U : I → C` of objects
+consists of the data of an element in `G.obj (Opposite.op T)` whenever we have elements
+`i` and `j` in `I` and maps `a : T ⟶ U i` and `b : T ⟶ U j`, and it must satisfy a compatibility
+with respect to precomposition. (When the binary product of `U i` and `U j` exists, this
+data for all `T`, `a` and `b` corresponds to the data of a section of `G` on this product.) -/
+@[ext]
+structure OneCochain where
+  /-- the data involved in a 1-cochain -/
+  ev (i j : I) ⦃T : C⦄ (a : T ⟶ U i) (b : T ⟶ U j) : G.obj (Opposite.op T)
+  ev_precomp (i j : I) ⦃T T' : C⦄ (φ : T ⟶ T') (a : T' ⟶ U i) (b : T' ⟶ U j) :
+    G.map φ.op (ev i j a b) = ev i j (φ ≫ a) (φ ≫ b) := by aesop
+
+namespace OneCochain
+
+attribute [simp] OneCochain.ev_precomp
+
+instance : One (OneCochain G U) where
+  one := { ev := fun _ _ _ _ _ => 1 }
+
+@[simp]
+lemma one_ev (i j : I) {T : C} (a : T ⟶ U i) (b : T ⟶ U j) :
+    (1 : OneCochain G U).ev i j a b = 1 := rfl
+
+variable {G U}
+
+instance : Mul (OneCochain G U) where
+  mul γ₁ γ₂ := { ev := fun i j T a b => γ₁.ev i j a b * γ₂.ev i j a b }
+
+@[simp]
+lemma mul_ev (γ₁ γ₂ : OneCochain G U) (i j : I) {T : C} (a : T ⟶ U i) (b : T ⟶ U j) :
+    (γ₁ * γ₂).ev i j a b = γ₁.ev i j a b * γ₂.ev i j a b := rfl
+
+instance : Inv (OneCochain G U) where
+  inv γ := { ev := fun i j T a b => (γ.ev i j a b) ⁻¹}
+
+@[simp]
+lemma inv_ev (γ : OneCochain G U) (i j : I) {T : C} (a : T ⟶ U i) (b : T ⟶ U j) :
+    (γ ⁻¹).ev i j a b = (γ.ev i j a b) ⁻¹ := rfl
```

Comment:
```suggestion
    (γ⁻¹).ev i j a b = (γ.ev i j a b)⁻¹ := rfl
```
Again I don't know the canonical style but I don't put these spaces in myself. If you want to leave them, and you know this is OK style, that's fine.